### PR TITLE
Pie fixes

### PR DESCRIPTION
**Pull request by bsdjhb**
This fixes the csu tests when PIE is enabled by default.  It also addresses a few mismerges where we kept local CheriBSD changes to add PIE support after merging in upstream PIE support resulting in some duplicate bits.

I haven't tested all of the makefile changes locally yet.

### CODE REVIEW

File: lib/csu/tests/fini_test.c [hunk 1]
```diff
@@ -141,7 +141,7 @@ dso_handle_check(void)
 {
 	void *dso = __dso_handle;
 
-#ifdef DSO_LIB
+#if defined(DSO_LIB) || defined(__PIE__)
 	ATF_REQUIRE_MSG(dso != NULL,
 	    "Null __dso_handle in DSO");
```
Comment (jrtc27) [inline lib/csu/tests/fini_test.c:146]:
```suggestion
	    "Null __dso_handle in DSO/PIE");
```
or
```suggestion
	    "Null __dso_handle in ET_DYN");
```

File: share/mk/bsd.prog.mk [hunk 2]
```diff
@@ -48,11 +48,15 @@ LDFLAGS+= -Wl,-znow
 .if ${MK_PIE} != "no"
 # Static PIE is not yet supported/tested.
 .if !defined(NO_SHARED) || ${NO_SHARED:tl} == "no"
+# PIE does not work for libcheri
+.if (defined(WANT_CHERI) && ${WANT_CHERI} == "sandbox") || \
+	(defined(LIBADD) && !empty(LIBADD:Mcheri))
```
Comment (jrtc27) [inline share/mk/bsd.prog.mk:53]: WANT_CHERI should be phased out. Checking LIBADD seems wrong; what about if a program indirectly links against libcheri?

Comment (jrtc27) [reply share/mk/bsd.prog.mk:53]: Though I see now this is just inherited from below...

Comment (bsdjhb) [reply share/mk/bsd.prog.mk:53]: Yeah, I'd actually rather just kill this and explicitly set MK_PIE=no in relevant Makefiles, but I'm not sure how many places I'd have to track down to make that work.

**Comment by zxombie** [comment]
Is is possible to build the csu test with pie disabled? The test should be testing `crtbegin.o` & `crtend.o` so we need to build in such a way as to use these rather than the `*S.o` versions.

**Comment by arichardson** [comment]
> Is is possible to build the csu test with pie disabled? The test should be testing `crtbegin.o` & `crtend.o` so we need to build in such a way as to use these rather than the `*S.o` versions.

Yeah I think ideally we would build both versions with -pie and -no-pie (or is it -nopie?).

File: lib/csu/tests/fini_test.c (hunk 1 above)
Comment (bsdjhb) [reply lib/csu/tests/fini_test.c]: Went with the first.

**Comment by bsdjhb** [comment]
This passes CI and now includes an explicit `dynamipie` test as well as forcing `dynamic` to be non-PIE.  The first two commits I have queued up to merge upstream to FreeBSD as well.

File: share/mk/bsd.prog.mk (hunk 2 above)
Comment (bsdjhb) [reply share/mk/bsd.prog.mk]: I've killed some of this (the WANT_CHERI bits) as it shouldn't be needed.

For libcheri we could explicitly set MK_PIE=no in the Makefiles of everything that links against libcheri which would let us remove the rest of this.  That seems a bit error-prone to do, but there are only about 10 places or so to set.

Comment (arichardson) [reply share/mk/bsd.prog.mk]: I think I prefer disabling pie for the libcheri bits. Shouldn't be that many and I think most might already be static so don't use pie?

**Review by arichardson — APPROVED** [review]

**Pull request merged**